x-pack/legacy/plugins/apm/public/components/app/Main/route_config/index.tsx

```diff
@@ -107,6 +111,74 @@ export const routes: BreadcrumbRoute[] = [
     ),
     name: RouteName.AGENT_CONFIGURATION
   },
+
+  {
+    exact: true,
+    path: '/settings/agent-configuration/create',
+    breadcrumb: i18n.translate(
+      'xpack.apm.breadcrumb.settings.agentConfigurationTitle',
+      { defaultMessage: 'Create Agent Configuration' }
+    ),
+    name: RouteName.AGENT_CONFIGURATION_CREATE,
+    component: () => {
+      const { search } = history.location;
+
+      // Ignoring here because we specifically DO NOT want to add the query params to the global route handler
+      // @ts-ignore
+      const { pageStep } = toQuery(search);
+
+      return (
+        <Settings>
+          <AgentConfigurationCreateEdit
+            pageStep={pageStep || 'choose-service-step'}
+          />
+        </Settings>
+      );
+    }
+  },
+  {
+    exact: true,
+    path: '/settings/agent-configuration/edit',
+    breadcrumb: i18n.translate(
+      'xpack.apm.breadcrumb.settings.agentConfigurationTitle',
+      { defaultMessage: 'Edit Agent Configuration' }
+    ),
+    name: RouteName.AGENT_CONFIGURATION_EDIT,
+    component: () => {
+      const { search } = history.location;
+
+      // typescript complains because `pageStop` does not exist in `APMQueryParams`
+      // Going forward we should move away from globally declared query params and this is a first step
+      // @ts-ignore
+      const { name, environment, pageStep } = toQuery(search);
+
+      // hooks rule is complaining about not using a hook here but it works so 🤷
+      // eslint-disable-next-line react-hooks/rules-of-hooks
+      const { data } = useFetcher(
+        callApmApi => {
+          return callApmApi({
+            method: 'POST',
+            pathname: '/api/apm/settings/agent-configuration/view',
+            params: { body: { service: { name, environment } } }
+          });
+        },
+        [name, environment]
+      );
```

Comment:
Controversy `#2`: fetching data in the route component. 

This makes the component a lot easier to test and work with, since it doesn't have any side effects. I think we should do this whereever possible.

Reply:
I like the approach, but I'm also worried about how big and complex this file will become.

Reply:
Data fetching is a side effect. By moving data feching out of the component (and into the route handler instead) the component only receives props.

Reply:
> but I'm also worried about how big and complex this file will become

yes, it could become a whole lot bigger than it is now. We may have to break it up like we've done on the server